shared/js/background/amp-protection.es6.js

```diff
@@ -0,0 +1,212 @@
+                return null
+            }
+
+            return newSite.url
+        }
+    }
+
+    return null
+}
+
+/**
+ * Check if the given request is a suspected 1st party AMP URL. If it is, then we will attempt to fetch the canonical URL
+ *
+ * @param {object} requestData - request data
+ * @param {object} thisTab - current tab
+ * @param {object} mainFrameRequestURL - main frame request url
+ * @returns true if the request is a suspected 1st party AMP url
+ */
+function tabNeedsDeepExtraction (requestData, thisTab, mainFrameRequestURL) {
+    if (utils.getBrowserName() !== 'moz') {
+        // deep extraction is only supported on Firefox
+        return false
+    }
+
+    if (requestIsExtension(requestData)) {
+        return false
+    }
+
+    if (thisTab.ampUrl && thisTab.ampUrl === mainFrameRequestURL.href) {
+        return false
+    }
+
+    return isAMPURL(mainFrameRequestURL.href)
+}
+
+/**
+ * This method checks if the given url is a 1st party AMP url
+ *
+ * @param {string} url - the url being loaded
+ * @returns true is the url is suspected to be a 1st party AMP url
+ */
+function isAMPURL (url) {
+    if (!ensureConfig()) {
+        return false
+    }
+
+    const site = new Site(url)
+    if (site.specialDomainName || !site.isFeatureEnabled(featureName)) {
+        return false
+    }
+
+    return ampSettings.keywords.some(keyword => url.includes(keyword))
+}
+
+/**
+ * This async method will fetch the DOM content of a suspected 1st party AMP URL and will
+ * return the canonical URL if one is found.
+ *
+ * @param {Site} site - the initiating site
+ * @param {string} url - the url being loaded
+ * @returns cancalon url if found, null otherwise
+ */
+async function fetchAMPURL (site, url) {
+    if (!ensureConfig()) {
+        return null
+    }
+
+    if (site.specialDomainName || !site.isFeatureEnabled(featureName) || site.url === 'about:blank') {
+        return null
+    }
+
+    let data = null
+    const timeoutController = new AbortController()
+    setTimeout(() => timeoutController.abort(), 5000)
+
+    try {
+        data = await fetch(url, { signal: timeoutController.signal })
+    } catch (e) {
+        console.error(e)
+        return null
+    }
+
+    if (!data) {
+        return null
+    }
+
+    const text = await data.text()
+    if (!text) {
+        return null
+    }
+
+    const doc = new DOMParser().parseFromString(text, 'text/html')
+    const errorNode = doc.querySelector('parsererror')
+    if (errorNode) {
+        // DOMNParser failed to parse the document
+        return null
+    }
+
+    const canonicalLinks = doc.querySelectorAll('[rel="canonical"]')
+
+    if (canonicalLinks.length > 0) {
+        const newSite = new Site(canonicalLinks[0].href)
+
+        if (newSite.specialDomainName || !newSite.isFeatureEnabled(featureName)) {
+            return null
+        }
+
+        return canonicalLinks[0].href
```

Comment:
Dropping 'All' from the method and renaming the variable reduces the overhead here a little:
```suggestion
    const firstCanonicalLink = doc.querySelector('[rel="canonical"]')

    if (firstCanonicalLink) {
        const newSite = new Site(firstCanonicalLink.href)

        if (newSite.specialDomainName || !newSite.isFeatureEnabled(featureName)) {
            return null
        }

        return firstCanonicalLink.href
```